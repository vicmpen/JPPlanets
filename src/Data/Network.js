import {NativeModules} from 'react-native';
const {NetworkManager} = NativeModules;

//function used internally to call the API
const get = async url => {
  console.log('----', NetworkManager);
  if (!url) {
    console.log('No url found, returning');
    return null;
  }

  let response = await NetworkManager.doGet(url);
  console.log(response);
  return response;
};
//Exported func to fetch the planets
const getAllPlanets = () => {
  return get('https://swapi.dev/api/planets/');
};

export default {
  getAllPlanets,
};
