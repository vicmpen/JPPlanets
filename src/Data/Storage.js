import AsyncStorage from '@react-native-async-storage/async-storage';

/**
 * Storage class to abstract away AsyncStorage. Generic get/set/delete item functions + some specific planet implementation
 * Works with stringified values.
 */

const saveItem = async (key, value) => {
  if (!key) {
    return false;
  }

  //currently we just accept string for saving
  if (typeof key != 'string') {
    throw createError('value should be a string');
  }

  try {
    await AsyncStorage.setItem(key, value);
    return true;
  } catch (e) {
    // saving error
    throw createError('error while saving value from storage');
  }
};

//
const getItem = async key => {
  if (!key) {
    return null;
  }
  try {
    const value = await AsyncStorage.getItem(key);
    //return the value or null, if not found
    return value;
  } catch (e) {
    // error reading value
    throw createError('error while reading value from storage');
  }
};

const createError = message => {
  return Error(message);
};

//Planet implementation of Storage
const storage_keys = {
  GET_ALL_PLANETS_KEY: 'ALL_PLANETS',
};

//TODO: Throw Error
const savePlanet = async (key, planets) => {
  try {
    let result = await saveItem(key, planets);
    return true;
  } catch (error) {
    console.log('Storage', error);
    return false;
  }
};

const saveAllPlanets = async planets => {
  console.log('Storage', 'saving planets', planets);
  return await saveItem(storage_keys.GET_ALL_PLANETS_KEY, planets);
};

//TODO: Throw Error
const getPlanet = async key => {
  try {
    let planets = await getItem(key);
    return planets;
  } catch (error) {
    return null;
  }
};

/**
 * Fetches all planets from the storage.
 * TODO: Throw Error
 */
const getAllPlanets = async () => {
  try {
    return await getItem(storage_keys.GET_ALL_PLANETS_KEY);
  } catch (error) {
    return createError('Error while Fetching Planets', error);
  }
};

const deleteAllPlanets = async () => {
  try {
    let _ = await AsyncStorage.removeItem(storage_keys.GET_ALL_PLANETS_KEY);
    console.log('Storage: Planets Deleted ');
  } catch (error) {
    return createError('Error while Deleting Planets', error);

  }
};

export default {
  saveAllPlanets,
  getAllPlanets,
  deleteAllPlanets,
};
