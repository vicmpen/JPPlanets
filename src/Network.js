import {NativeModules} from 'react-native';
const {NetworkManager} = NativeModules;

const get = async url => {
  console.log('----', NetworkManager);
  if (!url) {
    console.log('No url found, returning');
    return null;
  }

  let response = await NetworkManager.doGet(url);
  return JSON.parse(response);
};

export default {
  get,
};
