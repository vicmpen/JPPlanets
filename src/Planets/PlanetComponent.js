import React from 'react';
import {FlatList, Text, StyleSheet, View, TouchableOpacity} from 'react-native';
import PlanetRepository from '../Repositories/PlanetRepository';
import PlanetContainer, {usePlanets} from '../hooks/usePlanets';

const EmpyList = () => {
  return <Text>No Planets Nearby!</Text>;
};
//TODO: Add Referesh
const PlanetComponent = ({debug}) => {
  let [planets, loading] = usePlanets(false);
  return (
    <>
      <FlatList
        ListEmptyComponent={EmpyList}
        style={styles.flatlist}
        data={planets}
        renderItem={({item, index, separators}) => {
          return (
            <View
              style={{
                flex: 1,
                flexDirection: 'row',
                alignItems: 'center',
                justifyContent: 'center',
              }}>
              <Text
                style={{
                  flex: 1,
                  textAlign: 'center',
                }}>
                {item.name}
              </Text>
            </View>
          );
        }}
      />
      <PlanetRepoControlView debug={debug} />
    </>
  );
};

const styles = StyleSheet.create({
  flatlist: {
    flexGrow: 0,
    padding: 8,
    backgroundColor: 'yellow',
  },
});

export default PlanetComponent;

const PlanetRepoControlView = ({debug}) => {
  if (!debug) {
    return null;
  }
  return (
    <View>
      <TouchableOpacity onPress={PlanetRepository.deleteAllPlanets}>
        <Text>Clear Storage</Text>
      </TouchableOpacity>
    </View>
  );
};
