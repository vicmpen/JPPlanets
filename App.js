/**
 * Sample React Native App
 * https://github.com/facebook/react-native
 *
 * @format
 * @flow strict-local
 */

import React, {useState, useCallback, useEffect} from 'react';
import type {Node} from 'react';
import {
  SafeAreaView,
  StatusBar,
  useColorScheme,
  View,
  Text,
  FlatList,
  StyleSheet,
} from 'react-native';

import Network from './src/Network';
import {Colors, Header} from 'react-native/Libraries/NewAppScreen';

const App: () => Node = () => {
  let [planets, setPlanets] = useState([]);
  const isDarkMode = useColorScheme() === 'dark';

  const getPlanets = useCallback(async () => {
    let planets = await Network.get('https://swapi.dev/api/planets/');
    console.log('Planets Received!', planets);
    setPlanets(planets.results);
  }, []);

  // the useEffect is only there to call `someAsyncFunction` at the right time
  useEffect(() => {
    console.log('Calling getPlanets');
    getPlanets()
      // make sure to catch any error
      .catch(console.error);
  }, []);

  const backgroundStyle = {
    backgroundColor: isDarkMode ? Colors.darker : Colors.lighter,
  };
  return (
    <SafeAreaView style={styles.container}>
      <StatusBar barStyle={isDarkMode ? 'light-content' : 'dark-content'} />
      <FlatList
        style={styles.flatlist}
        data={planets}
        renderItem={({item, index, separators}) => {
          console.log('item!', item);
          return <Text style={{flex: 1}}>{item.name}</Text>;
        }}
      />
    </SafeAreaView>
  );
};

export default App;

const styles = StyleSheet.create({
  container: {
    flexDirection: 'column',
    flex: 1,
    backgroundColor: 'black',
  },
  flatlist: {
    flex: 1,
    padding: 8,
    backgroundColor: 'yellow',
  },
});
