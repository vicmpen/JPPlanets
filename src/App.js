/**
 * Sample React Native App
 * https://github.com/facebook/react-native
 *
 * @format
 * @flow strict-local
 */

import x from './Polyfills/console';

import React, {useState, useCallback, useEffect} from 'react';

import {
  SafeAreaView,
  StatusBar,
  useColorScheme,
  StyleSheet,
  View,
  Text,
} from 'react-native';

import PlanetRepository from './Repositories/PlanetRepository';
import PlanetList from './Planets/PlanetComponent';
import PlanetComponent from './Planets/PlanetComponent';
import AppConfig from './AppConfig';
import {useNetInfo} from '@react-native-community/netinfo';

const App = () => {
  const isDarkMode = useColorScheme() === 'dark';
  return (
    <SafeAreaView style={styles.container}>
      <StatusBar barStyle={isDarkMode ? 'light-content' : 'dark-content'} />
      <NetInfo />
      <PlanetComponent debug={AppConfig.debug_mode} />
    </SafeAreaView>
  );
};

export default App;

const styles = StyleSheet.create({
  container: {
    flexDirection: 'column',
    flex: 1,
    backgroundColor: 'white',
  },
  flatlist: {
    flexGrow: 0,
    padding: 8,
    backgroundColor: 'yellow',
  },
});

export const NetInfo = () => {
  return (
    <View style={{height: 50, width: '100%', backgroundColor: 'green'}}>
      <Text>Device Online: {useNetInfo().isConnected?.toString()}</Text>
    </View>
  );
};
