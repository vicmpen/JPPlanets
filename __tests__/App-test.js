/**
 * @format
 */

import 'react-native';
import React from 'react';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
import App from '../src/App';
import Storage from '../src/Data/Storage';
import AsyncStorage from '@react-native-async-storage/async-storage';
import Network from '../src/Data/Network';
import PlanetComponent from '../src/Planets/PlanetComponent';

jest.mock('@react-native-async-storage/async-storage', () =>
  require('@react-native-async-storage/async-storage/jest/async-storage-mock'),
);

describe('Storage Module Mocking', () => {
  it('checks if Storage Saves Correctly', async () => {
    let payload = JSON.stringify(['planet1', 'planet2']);
    await Storage.saveAllPlanets(payload);
    expect(AsyncStorage.setItem).toBeCalledWith('ALL_PLANETS', payload);
  });

  it('checks if Storage Deletes Correctly', async () => {
    await Storage.deleteAllPlanets();
    expect(AsyncStorage.removeItem).toBeCalledWith('ALL_PLANETS');
  });
});
