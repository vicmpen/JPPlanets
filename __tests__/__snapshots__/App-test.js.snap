// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storage Module Mocking renders correctly 1`] = `
Array [
  <RCTScrollView
    ListEmptyComponent={[Function]}
    data={Array []}
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    style={
      Object {
        "backgroundColor": "yellow",
        "flexGrow": 0,
        "padding": 8,
      }
    }
    viewabilityConfigCallbackPairs={Array []}
  >
    <View>
      <Text>
        No Planets Nearby!
      </Text>
    </View>
  </RCTScrollView>,
  <View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <Text>
        Clear Storage
      </Text>
    </View>
  </View>,
]
`;
